core/src/test/java/org/infinispan/persistence/file/SingleFileGracefulShutdownMigrationTest.java

```diff
@@ -0,0 +1,104 @@
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.marshall.TestObjectStreamMarshaller;
+import org.infinispan.test.AbstractInfinispanTest;
+import org.infinispan.test.TestDataSCIImpl;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+/**
+ * SFS content created on 11.0.x branch using the below code
+ * <p>
+ * sfs/corrupt/all-entries.dat contains all off below entries sfs/corrupt/no-private-metadata.dat does not include the
+ * final two entries with PrivateMetadata sfs/corrupt/private-metadata-only.dat only contains the final two entries with
+ * PrivateMetadata
+ * <p>
+ * These files were loaded using 12.1.x SFS so that they were migrated with the broken code. Hence, when the .dat files
+ * are loaded in this test the SFS attempts to recover the corrupt data.
+ *
+ * @author Ryan Emerson
+ * @since 13.0
+ */
+@Test(groups = "unit", testName = "persistence.file.SingleFileGracefulShutdownMigrationTest")
+public class SingleFileGracefulShutdownMigrationTest extends AbstractInfinispanTest {
+
+   private static final String CACHE_NAME = "update-cache";
+
+   private String tmpDirectory;
+
+   @BeforeClass
+   protected void setUpTempDir() throws IOException {
+      tmpDirectory = CommonsTestingUtil.tmpDirectory(this.getClass());
+      new File(tmpDirectory).mkdirs();
+   }
+
+   @AfterClass
+   protected void clearTempDir() {
+      Util.recursiveFileRemove(tmpDirectory);
+   }
+
+   @DataProvider(name = "testFiles")
+   Object[][] singleTypes() {
+      return new Object[][] {
+            {"sfs/corrupt/all-entries.dat"},
+            {"sfs/corrupt/private-metadata-only.dat"},
+            {"sfs/corrupt/no-private-metadata.dat"},
+            {"sfs/11_0/sfs-store-cache.dat"},
+            {"sfs/12_0/all-entries.dat"},
+            {"sfs/12_1/all-entries.dat"},
+      };
+   }
+
+   @Test(dataProvider = "testFiles")
+   public void testAllEntriesRecovered(String fileName) throws Exception {
+      InputStream is = FileLookupFactory.newInstance().lookupFile(fileName, Thread.currentThread().getContextClassLoader());
+      Files.copy(is, Paths.get(tmpDirectory).resolve(CACHE_NAME + ".dat"), StandardCopyOption.REPLACE_EXISTING);
+
+      ConfigurationBuilderHolder cbh = new ConfigurationBuilderHolder();
+      cbh.getGlobalConfigurationBuilder().serialization().addContextInitializer(new TestDataSCIImpl());
+      cbh.newConfigurationBuilder(CACHE_NAME)
+            .persistence()
+            .addSingleFileStore()
+            .segmented(false)
+            .location(tmpDirectory);
+
+      try (EmbeddedCacheManager cacheManager = new DefaultCacheManager(cbh, true)) {
+         Cache<Object, Object> cache = cacheManager.getCache(CACHE_NAME);
+         // Iterate all entries to ensure values can be read
+         cache.forEach((k, v) -> {
```

Comment:
I'm not sure what this tests. 
Isn't possible if there is a failure reading, the `Consumer` won't be invoked? 
Can't we just add a counter to the `Consumer` and check if it has seen the correct number of entries?

Reply:
> Isn't possible if there is a failure reading, the `Consumer` won't be invoked?

Yes, but the test will fail in this case.

> Can't we just add a counter to the `Consumer` and check if it has seen the correct number of entries?

The `sfs/corrupt/all-entries.dat` file will actually result in `2003` entries, opposed to `1003` for the other tests, as we're resetting the lost `created` and `lastUsed` values to the current wall time when recovering the entries. So `assertNotNull` was simpler at the time. We could add a third arg to the test so we have something like:

```
            {"sfs/corrupt/all-entries.dat", Marshaller.PROTOSTREAM, 2003},
            {"sfs/12_0/all-entries.dat", Marshaller.PROTOSTREAM, 1003},
```

Reply:
this looks good too me :)
```
            {"sfs/corrupt/all-entries.dat", Marshaller.PROTOSTREAM, 2003},
            {"sfs/12_0/all-entries.dat", Marshaller.PROTOSTREAM, 1003},
```